oximeter/producer/src/lib.rs

```diff
@@ -274,21 +423,152 @@ pub async fn register(
     }
 }
 
-/// Handle a request to pull available metric data from a [`ProducerRegistry`].
-pub async fn collect(
-    registry: &ProducerRegistry,
-    producer_id: Uuid,
-) -> Result<HttpResponseOk<ProducerResults>, HttpError> {
-    if producer_id == registry.producer_id() {
-        Ok(HttpResponseOk(registry.collect()))
-    } else {
-        Err(HttpError::for_not_found(
-            None,
-            format!(
-                "Producer ID {} is not valid, expected {}",
-                producer_id,
-                registry.producer_id()
+/// Resolve Nexus's address using the provided resolver.
+async fn resolve_nexus_with_backoff(
+    log: &Logger,
+    resolver: &Resolver,
+) -> SocketAddr {
+    let log_failure = |error, delay| {
+        warn!(
+            log,
+            "failed to lookup Nexus IP, will retry";
+            "delay" => ?delay,
+            "error" => ?error,
+        );
+    };
+    let do_lookup = || async {
+        resolver
+            .lookup_socket_v6(ServiceName::Nexus)
+            .await
+            .map_err(|e| BackoffError::transient(e.to_string()))
+            .map(Into::into)
+    };
+    backoff::retry_notify(
+        backoff::retry_policy_internal_service(),
+        do_lookup,
+        log_failure,
+    )
+    .await
+    .expect("Expected infinite retry loop resolving Nexus address")
+}
+
+/// Register as a metric producer with Nexus, retrying endlessly with backoff.
+///
+/// This returns the lease renewal period that we're required to re-register
+/// within.
+async fn register_with_backoff(
+    addr: SocketAddr,
+    log: &Logger,
+    endpoint: &ApiProducerEndpoint,
+) -> Duration {
+    let log_failure = |error, delay| {
+        warn!(
+            log,
+            "failed to register as a producer with Nexus, will retry";
+            "delay" => ?delay,
+            "error" => ?error,
+        );
+    };
+    let do_register = || async {
+        register_once(addr, log, endpoint).await.map_err(|e| match e {
+            Error::RegistrationError { retryable, msg } => {
+                if retryable {
+                    BackoffError::transient(msg)
+                } else {
+                    BackoffError::permanent(msg)
+                }
+            }
+            _ => BackoffError::permanent(e.to_string()),
+        })
+    };
+    backoff::retry_notify(
+        backoff::retry_policy_internal_service(),
+        do_register,
+        log_failure,
+    )
+    .await
+    .expect("Expected infinite retry loop registering as a producer with")
```

Comment:
I don't think we can `.expect` here - the two `BackoffError::permanent` points above mean it _isn't_ an infinite retry loop. I'm not sure what we're supposed to do with a permanent registration error, though?

Reply:
Good point, this is not an infinite loop any more. And I'm also not sure what to do with a permanent error. I mean, what _can_ we do? I think the only things are (1) panicking, not great; or (2) logging it, and stopping the registration loop. If we do that, is it worth adding methods for fetching the registration state, such as the result of the last registration (and timestamp)? I don't know how we'd use it, but adding a way to kick the registration task to try again might be useful too.

Reply:
I think the two `_ =>`s in this call chain make me a little nervous (one on line 481, and one on line 418). Even if all the variants covered by `_` today really are permanent errors, will that be true for new error variants added in the future?

I'm moderately tempted to suggest we remove the permanent errors altogether here and make all errors transient. E.g., if we get some response from Nexus that indicates a permanent error, but it came from a Nexus that's borked, could we succeed if we resolve to a different Nexus?

Reply:
That's a good question. I wrote it this way a long time ago to avoid retrying operations that could never succeed, e.g., making the same bad request to Nexus forever. I get your point though, that the "same request" may actually succeed in the future, e.g., when we support online updates of Nexus or blue-green deployments in which we resolve to a new Nexus that _can_ process the request for us.

It does make me a bit nervous to continually try bad requests. I'm worried about the availability hit we might take, really. But I agree that there's nothing obviously better to do in this case, and backoff should protect us from the worst of that.

Reply:
Ok, this is done in 1413ce8ed. I've also removed the `Error::RegistrationError` variant entirely, since it was only used internally, to communicate if we should retry a registration request. We always do that now, so no need for the variant at all.